README.md
# The `emoji` package

[![CTAN](https://img.shields.io/ctan/v/emoji.svg)](https://www.ctan.org/pkg/emoji)
[![GitHub release](https://img.shields.io/github/release/stone-zeng/latex-emoji/all.svg)](https://github.com/stone-zeng/latex-emoji/releases/latest)

Emoji support in (Lua)LaTeX.

## Introduction

The `emoji` package allows user to typeset emoji in a LaTeX document. It requires LuaHBTeX engine,
which can be called by `lualatex` in TeX Live 2020 or `lualatex-dev` in TeX Live 2019.

## Usage

```tex
\documentclass{article}
\usepackage{emoji}
\setemojifont{Apple Color Emoji}  % Optional

\begin{document}
\emoji{joy}
\emoji{+1}
\emoji{family-man-woman-girl-boy}
\emoji{woman-and-man-holding-hands-light-skin-tone-medium-dark-skin-tone}
\end{document}
```

Result:

> &#x1F602; &#x1F44D; &#x1F468;&#x200D;&#x1F469;&#x200D;&#x1F467;&#x200D;&#x1F466; &#x1F469;&#x1F3FB;&#x200D;&#x1F91D;&#x200D;&#x1F468;&#x1F3FE;

## License

This work may be distributed and/or modified under the conditions of the [LaTeX Project Public License](http://www.latex-project.org/lppl.txt), either version 1.3c of this license or (at your option) any later version.

-----

Copyright (C) 2020 by Xiangdong Zeng.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/).

## [Unreleased]

## [v0.2.2] - 2021-07-17

### Added

- Support Unicode 13.1 - [#5](https://github.com/stone-zeng/latex-emoji/issues/5)

### Fixed

- Compatibility with `luatexja` - [#4](https://github.com/stone-zeng/latex-emoji/issues/4)
- Fix incorrect aliases - [#7](https://github.com/stone-zeng/latex-emoji/issues/7)

## [v0.2.1] - 2020-06-27

### Fixed

- Update documentation to fully support Unicode 12.1

## [v0.2] - 2020-03-16

### Added

- Add separated `emoji-table.def`
- Support Unicode 13.0
- Detect Noto Color Emoji automatically

### Fixed

- Fix preamble detection
- Update documentation for history of LuaHBTeX

## [v0.1] - 2020-02-23

### Added

- First (experimental) version of the `emoji` package

[Unreleased]: https://github.com/stone-zeng/latex-emoji/compare/v0.2.2...HEAD
[v0.2.2]: https://github.com/stone-zeng/latex-emoji/compare/v0.2.1...v0.2.2
[v0.2.1]: https://github.com/stone-zeng/latex-emoji/compare/v0.2...v0.2.1
[v0.2]: https://github.com/stone-zeng/latex-emoji/compare/v0.1...v0.2
[v0.1]: https://github.com/stone-zeng/latex-emoji/releases/tag/v0.1


release.sh
#!/bin/sh

# latexmk -pdflua emoji-doc.tex

working_dir=$PWD
tmp_dir=/tmp/emoji
tds_dir=$tmp_dir/tds
ctan_dir=$tmp_dir/emoji
tex_dir=$tds_dir/tex/latex/emoji
doc_dir=$tds_dir/doc/latex/emoji

mkdir -p $tmp_dir
mkdir -p $tds_dir
mkdir -p $ctan_dir
mkdir -p $tex_dir
mkdir -p $doc_dir

cp emoji.sty       $ctan_dir
cp emoji-table.def $ctan_dir
cp emoji-doc.tex   $ctan_dir
cp emoji-doc.pdf   $ctan_dir
cp README.md       $ctan_dir

chmod 644 $ctan_dir/*.*

cp $ctan_dir/emoji.sty       $tex_dir
cp $ctan_dir/emoji-table.def $tex_dir
cp $ctan_dir/emoji-doc.tex   $doc_dir
cp $ctan_dir/emoji-doc.pdf   $doc_dir
cp $ctan_dir/README.md       $doc_dir

cd $tds_dir
zip -q -r -9 emoji.tds.zip .

cp $tds_dir/*.zip $tmp_dir
rm -r $tds_dir

cd $tmp_dir
zip -q -r -9 emoji.zip .

cd $working_dir
mkdir -p ./release
cp -f $tmp_dir/emoji.zip     ./release
cp -f $tmp_dir/emoji.tds.zip ./release

rm -r $tmp_dir


data/build.py
# -*- coding: utf-8 -*-

import json
import re
import unicodedata as ud

COPYRIGHT_NOTICE = r'''%%
%%  Copyright (C) 2020, 2021 by Xiangdong Zeng <xdzeng96@gmail.com>
%%
%%  This work may be distributed and/or modified under the
%%  conditions of the LaTeX Project Public License, either
%%  version 1.3c of this license or (at your option) any later
%%  version. The latest version of this license is in:
%%
%%    http://www.latex-project.org/lppl.txt
%%
%%  and version 1.3 or later is part of all distributions of
%%  LaTeX version 2005/12/01 or later.
%%
%%  This work has the LPPL maintenance status `maintained'.
%%
%%  The Current Maintainer of this work is Xiangdong Zeng.
%%
'''

UNICODE_EMOJI_DATA_FILE = './data/emoji-test.txt'
GITHUB_EMOJI_DATA_FILE = './data/emoji.json'

OUTPUT_FILE = 'emoji-table.def'
OUTPUT_FILE_INFO = '{2021/07/17}{0.2.2}{Emoji support in (Lua)LaTeX}'

# 0 = code points, 1 = version, 2 = description
EMOJI_ENTRY_PATTERN = re.compile(r'^(.+?)\W+;\W(?:fully-qualified|component).+E(\d+\.\d+)\W+(.+)')

def to_code_points(s: str):
    return ' '.join([hex(ord(i))[2:].upper() for i in s])

def normalize_name(s: str):
    # Remove accents
    s = ''.join(c for c in ud.normalize('NFD', s.lower()) if ud.category(c) != 'Mn')
    s = re.sub(r',|\.|:|\(|\)|’|“|”|&', '', s)
    s = re.sub(r'\s+', '-', s)
    return s.replace('#', 'hash').replace('*', 'asterisk')

def normalize_aliases(aliases: list[str], names: set[str], previous_aliases: set[str]):
    return sorted(set(s.replace('_', '-') for s in aliases) - (names | previous_aliases))

def to_tex_code_points(c_list: list[str]):
    return ''.join(to_tex_hex(c) for c in c_list).lower()

def to_tex_hex(s: str):
    if len(s) == 4:
        return '^^^^' + s
    if len(s) == 5:
        return '^^^^^^0' + s

def texify(s: str):
    replace = {'&': '\\&', '#': '\\#', '’': '\'', '“': '``', '”': '\'\'', ' ': '~'}
    for k, v in replace.items():
        s = s.replace(k, v)
    return s[0].capitalize() + s[1:]

# Read from Unicode data
with open(UNICODE_EMOJI_DATA_FILE) as f:
    emoji = {}
    for line in f.readlines():
        if line.startswith('# group'):
            group_name = line[9:-1]
            emoji[group_name] = {}
        elif line.startswith('# subgroup'):
            subgroup_name = line[12:-1]
            emoji[group_name][subgroup_name] = []
        elif not line.startswith('#') and line != '\n':
            entry = re.findall(EMOJI_ENTRY_PATTERN, line)
            if entry:
                emoji[group_name][subgroup_name].append(entry[0])

# Read from GitHub data
with open(GITHUB_EMOJI_DATA_FILE) as f:
    github_data = {}
    github_hex_data = {}
    for i in json.load(f):
        github_data[i['description']] = i['aliases']
        github_hex_data[to_code_points(i['emoji'])] = i['aliases']

# Normalize emoji data
emoji_names = set()
for i in emoji:
    for j in emoji[i]:
        entries = []
        for k in emoji[i][j]:
            (code_points, version, description) = (k[0], k[1], k[2])
            name = normalize_name(description)
            aliases = github_hex_data.get(code_points, []) + github_data.get(description, [])
            emoji_names.add(name)
            entries.append({
                'code_points': to_tex_code_points(code_points.split()),
                'name': name,
                'aliases': aliases,
                'version': version,
                'description': description,
            })
        emoji[i][j] = entries

# Remove duplicated aliases
# The aliases should be distinct from all the emoji names and previous aliases
previous_aliases = set()
for i in emoji:
    for j in emoji[i]:
        for k in emoji[i][j]:
            aliases = normalize_aliases(k['aliases'], emoji_names, previous_aliases)
            previous_aliases.update(aliases)
            k['aliases'] = ', '.join(aliases)

# Write into LaTeX file
with open(OUTPUT_FILE, 'w') as f:
    f.write(COPYRIGHT_NOTICE)
    f.write(f'\\ProvidesExplFile{{{OUTPUT_FILE}}}\n')
    f.write(f'  {OUTPUT_FILE_INFO}\n')
    for i in emoji:
        f.write(f'\\__emoji_group:n {{{texify(i)}}}\n')
        for j in emoji[i]:
            f.write(f'\\__emoji_subgroup:n {{{texify(j)}}}\n')
            for k in emoji[i][j]:
                line = ('\\__emoji_def:nnnnn' + ' {{{}}}' * 5).format(
                    k['code_points'],
                    k['name'],
                    k['aliases'],
                    texify(k['description']),
                    k['version']
                )
                f.write(line + '\n')
    f.write(f'%%\n%% End of file `{OUTPUT_FILE}\'.\n')
